File: 020-valid-parentheses/valid-parentheses.js
// Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
//
// The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.


/**
 * @param {string} s
 * @return {boolean}
 */
// a valid one should be like: '({[]})'
// in the stack -> [')', '}', ']']
var isValid = function(s) {
    let res = []
    let charArr = s.split('')
    const dict = {
        '(': ')',
        '[': ']',
        '{': '}'
    }
    
    for (let p of charArr) {
        if (p === '(' || p === '{' || p === '[') {
            res.push(dict[p])
        } else {
            let curr = res.pop()
            if (p !== curr) return false
        }
    }
    return res.length === 0
};

File: 020-valid-parentheses/valid-parentheses.py
# -*- coding:utf-8 -*-


# Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
#
# The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s) % 2 != 0:
            return False
        dict = {'}': '{', ')': '(', ']': '['}
        stack = []
        
        for p in s:
            if p in ['(', '{', '[']:
                stack.append(p)
            else:
                if len(stack) == 0 or stack.pop() != dict[p]:
                    return False
        return len(stack) == 0
    
